### DISCUSSION TITLE
Question about SAML

### DISCUSSION THREAD

**Discussion by helmut72** [zitadel/zitadel#4635] [Q&A] [answered]
Hi,

I want to use Zitadel to authenticate in [Grist](https://getgrist.com). I have a working configuration with Keycloak, used the info from here:
https://github.com/gristlabs/grist-core/issues/44#issuecomment-1043709852

I read through the existing SAML examples for Zitadel. For example the one for Gitlab:
https://docs.zitadel.com/docs/guides/integrate/gitlab-saml

Also read here ...
https://docs.zitadel.com/docs/apis/saml/endpoints#saml-20-metadata

... and here for Grist:
https://support.getgrist.com/install/saml/

Problem/Question 1:
From Keycloak I got the key/cert for ENV `GRIST_SAML_SP_KEY` and `GRIST_SAML_SP_KEY`.

Is this something that also exist in Zitadel (to download) or do I need create a SP key/cert for myself? Are there some specific requirements for Zitadel?

For now I use the one I got from Keycloak, because openssl doesn't show any Keycloak specific stuff. Public Key Algorithm is rsaEncryption and Signature Algorithm is sha256WithRSAEncryption.

Problem/Question 2, 3 and 4:
For `GRIST_SAML_IDP_LOGIN` and `GRIST_SAML_IDP_LOGIN` I use `https://zitadel.example.com/saml/v2/auth`. When opening `https://grist.example.com, browser will be forwarded to an URL like `https://zitadel.example.com/saml/v2/SSO?SAMLRequest=hVJLT...` and stops with a XML Page and the status message `failed to verify signature: crypto/rsa: verification error`.

Have tried combinations in Grist with `GRIST_SAML_IDP_UNENCRYPTED=0/1` and in Zitadel adding/deleting ENV `ZITADEL_SAML_IDPCONFIG_WANTAUTHREQUESTSIGNED=false`, but I always got the same message.

Are the login/logout URLs ok for Zitadel?
Are there some more config settings in Zitadel?
How can I get a more verbose log? ENV `ZITADEL_LOG_LEVEL=debug` is enabled.

Thank you.

**Comment by fforootd** [comment]
@stebenz can you have a look into this?

**Comment by stebenz** [comment]
Hello @helmut72 

I try to cover all your questions, if something is unclear or I missed something just mention it.

Question 1:
If you need any certificates that Grist should use, then you would have to generate them yourself, if you mean the certificates which are used by ZITADEL for signing they are included in https://zitadel.example.com/saml/v2/metadata or specifically for signing the responses https://zitadel.example.com/saml/v2/metadata .

Question 2:
GRIST_SAML_IDP_LOGIN should be https://zitadel.example.com/saml/v2/SSO as it is the standpoint for the login process, https://zitadel.example.com/saml/v2/SLO for logout.

Question 3:
GRIST_SAML_IDP_UNENCRYPTED, I'm not sure if this results in an unsigned request, if you could provide me with an example request I could have a look at it. But there are some more config settings in ZITADEL, but we can have a look at them as necessary.

Question 4:
ZITADEL_LOG_LEVEL=debug should provide you with a verbose log, is there anything else you desire as additional information you want to see in the logs?

**Reply by helmut72** [reply]
Hi @stebenz 

thank you!

> GRIST_SAML_IDP_LOGIN should be https://zitadel.example.com/saml/v2/SSO as it is the standpoint for the login process, https://zitadel.example.com/saml/v2/SLO for logout.

I have changed it, but the result is the same. ZITADEL just shows this output:
![saml-zitadel](https://user-images.githubusercontent.com/1388751/200672391-04b964c1-7b27-4899-9b27-251b58bdf5ba.jpg)

> GRIST_SAML_IDP_UNENCRYPTED, I'm not sure if this results in an unsigned request, if you could provide me with an example request I could have a look at it.

It's a request like this:

```
https://zitadel.example.com/saml/v2/SSO?
    SAMLRequest=<long-string>&
    RelayState=permit-external-aaaaaaaa-xxxx-yyyy-zzzz-123456789012&
    SigAlg=http%3A%2F%2Fwww.w3.org%2F2001%2F04%2Fxmldsig-more%23rsa-sha256&
    Signature=<long-string>
```

> But there are some more config settings in ZITADEL

Is there somewhere a documentation for SAML?

> is there anything else you desire as additional information you want to see in the logs?

Maybe I have wrong expectations, but just look what Apache or Caddy provides when they are set into debug mode. When trying Grist with ZITADEL, ZITADEL only output this line:
```
time="2022-11-08T20:39:48Z" level=error msg="crypto/rsa: verification error" caller="/home/runner/go/pkg/mod/github.com/zitadel/saml@v0.0.8/pkg/provider/checker/checker.go:98"
```

And it looks like the message "crypto/rsa: verification error" is also not enough for you... ;) ... it would be great if there are some more meaningful messages like "verification error, because ...".

Thanks!

**Reply by stebenz** [reply]
Hi @helmut72 

No worries, here to help.
First off, it's unfortunate that I could only try it myself with Grist in a custom pricing, so I hope you bear with me here.

It seems like GRIST_SAML_IDP_UNENCRYPTED means that the request is still signed, as a Signature is included.
The output tells me that ZITADEL couldn't validate the provided signature with the saved metadata in the SAML application on our side.
Could you do me a favor and try to validate the signature with for example https://www.samltool.com/validate_authn_req.php where you just have to fill in the SAMLRequest, SP entityID(from the metadata of grist), target URL(https://zitadel.example.com/saml/v2/SSO), X509 cert(from grist), signature and the SigAlg?

There is currently only basic documentation for SAML, but there is also not that much configurable, mostly provided information like contact for example, which is shown in the metadata, or custom endpoints. (default defined https://github.com/zitadel/zitadel/blob/main/cmd/defaults.yaml#L212). As a lot of the communication is given through metadata sharing.

True, I planned to look into improving the logging for SAML sometime soon, thanks.

**Reply by helmut72** [reply]
Hi @stebenz 

thanks for your support!

> it's unfortunate that I could only try it myself with Grist in a custom pricing

I also don't use Grists cloud service. It's self hosted with docker-compose.

> Could you do me a favor and try to validate the signature with for example...

Without "Private Key of the Identity Provider" and "Signature of the SAML AuthN Request" it outputs:

<img width="398" alt="Bildschirm­foto 2022-11-09 um 21 38 06" src="https://user-images.githubusercontent.com/1388751/200937866-04872468-76d6-482f-9633-59dbaa6e1f90.png">

> (default defined https://github.com/zitadel/zitadel/blob/main/cmd/defaults.yaml#L212).

Is it enough to only set `ZITADEL_SAML_IDPCONFIG_WANTAUTHREQUESTSIGNED=false` to enable or disable this option or do I need to set everything under the SAML context:
https://github.com/zitadel/zitadel/blob/main/cmd/defaults.yaml#L212-L227
?

Thank you.

**Reply by stebenz** [reply]
> I also don't use Grists cloud service. It's self hosted with docker-compose.

Ah nice, didn't see that there is a possibility to deploy it locally, maybe will look into that.

> Without "Private Key of the Identity Provider" and "Signature of the SAML AuthN Request" it outputs:

Hm but if you leave auth the Signature, there is nothing to validate, exactly there the error occurs, what do you get if you include the Signature? Private key is only necessary if the message is encrypted, which it is hopefully not.

> Is it enough to only set `ZITADEL_SAML_IDPCONFIG_WANTAUTHREQUESTSIGNED=false` to enable or disable this option or do I need to set everything under the SAML context: https://github.com/zitadel/zitadel/blob/main/cmd/defaults.yaml#L212-L227 ?

Yes, the other information should have no impact currently.
The flag "WantAuthRequestSigned" should just tell the serviceprovider that the request has to be signed, where as there is a flag for the SP "AuthnRequestsSigned" which does the same on the other side, included in the metadata on each side.

**Reply by helmut72** [reply]
> Yes, the other information should have no impact currently.
> The flag "WantAuthRequestSigned" should just tell the serviceprovider that the request has to be signed

Thank you. But it doesn't matter if I set it to true or false, I always get just this XML output.

**Reply by stebenz** [reply]
Hi @helmut72

Ok sooooo, I did try it out locally and it worked without any errors, to confirm the steps with you here is how I did it:

- used the Auth0 example as base [here](https://support.getgrist.com/install/saml/#example-auth0)
- created PEM and Key for Grist like [here](https://auth0.com/docs/authenticate/protocols/saml/saml-sso-integrations/sign-and-encrypt-saml-requests#use-a-custom-key-to-sign-requests)
- Download the certificate from ZITADEL with URL https://zitadel.exmaple.com/saml/v2/certificate
- Copied these 3 files (zitadel.cert, girst.cert, grist.key) into a local folder, for me, it was $HOME/grist
- Started Grist as docker with the following command: 
```docker run -p 8484:8484 \                                                                                                                                                                            
  -v $HOME/grist:/persist \
  -e GRIST_SESSION_SECRET=invent-a-secret-here \
  -e GRIST_SAML_SP_HOST='http://localhost:8484' \
  -e GRIST_SAML_IDP_UNENCRYPTED=1 \
  -e GRIST_SAML_IDP_LOGIN='http://example.zitadel.com/saml/v2/SSO' \
  -e GRIST_SAML_IDP_LOGOUT='http://example.zitadel.com/saml/v2/SLO' \
  -e GRIST_SAML_IDP_CERTS='/persist/zitadel.cert' \
  -e GRIST_SAML_SP_KEY='/persist/grist.key' \
  -e GRIST_SAML_SP_CERT='/persist/grist.cert' \
  -it gristlabs/grist
```
- Added SAML application in ZITADEL with a unique name and metadata from "http://localhost:8484/saml/metadata.xml", whereas you can also look at the grist metadata under this URL

For me, it was "localhost:8080" as I ran it locally, but I replaced it with "example.zitadel.com" to hold on to our start scenario.

Can you try it like so? 
Or can you tell me what of this process you did differently? 
So we can find out if there is still a problem or if something was missing from your setup?

**Reply by helmut72** [reply]
Hi @stebenz 

thank you for trying it out!

It still didn't work on my installation. My configuration looks nearly the same like yours but only with some more configuration because I use a reverse proxy in front of Grist. This doesn't matter for saml.

But I now know the cause why it doesn't work for me. It works now!

On my first many try and error I always pasted Grists metadata url into "Url where Metadata file is located". Never veryfied if there are changes, but better safe than sorry. Because Zitadel only shows something like "no changes", I don't tried to do that later.

Now, started again from scratch with Grist and you showed me that it works, I deleted Zitadels saml config for Grist completely and added it again. That was the solution! Looks like no metadata xml url will be loaded any 2nd time in Zitadel. For me this looks like a bug!? I have double checked. After it worked, I switched back to my sp certs from Keycloak, got the same error, try to load new metadata url in Zitadel, still doesn't work, removed and added saml config for Grist in Zitadel again: works

But now the next problem. There isn't the complete user name and mail addres transmitted from Zitadel to Grist. Grist mentions:
```
We expect IdP to provide us with name_id, a unique identifier for the user. We also use optional attributes for the user’s name, for which we accept any of:

FirstName
LastName
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname
You may need to tweak your IdP’s defaults to match Grist’s expectations.
```

Also from the link in my first post, in Keycloak it is required to do:
```
Protocol mappers are needed:
givenName: http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname
surname: http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname
email: http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress
```

Edit: Continued playing with Grist and having no valid mail address is a showstopper using Zitadel with Grist, because Grist requires a valid mail address, eg. for sharing documents with family members.

**Reply by stebenz** [reply]
Hi @helmut72 

Yes, the functionality to provide an URL instead of an entire metadata file is at the moment only for convenience, there is no reload implemented yet.

So it seems as if grist needs attributes with defined names and the option to receive the name_id with a specific type.
We planned to include this functionality, maybe into actions but were not quite sure how the need would be.

Do you have any timeframe in which you would need this to work?

I would be happy to open up an issue regarding this functionality, I would also be happy to bring it up if you wanted to open the issue yourself. 
The problem is on our side, that most of the applications which provide SAML have the possibility to configure which attributes are used for what on their side, so it was not our first priority.

**Comment by helmut72** [comment]
Hi @stebenz 

> Yes, the functionality to provide an URL instead of an entire metadata file is at the moment only for convenience, there is no reload implemented yet.

It's a great convenience when it works in the wizard. When you re-enter the SAML configuration it's still there, but doesn't work. Don't you think it must be hidden or grayed out then, because otherwise it's an error?

> Do you have any timeframe in which you would need this to work?

No serious timeframe but I want to finish my private homelab in spring 2023. After a lot of research in the last two years. To continue my real project being based on my homelab. For now I just test out ZITADEL against other IDPs.

> The problem is on our side, that most of the applications which provide SAML have the possibility to configure which attributes are used for what on their side, so it was not our first priority.

Don't know what features an IDP and what a SP must at least provide, that both work together.

I'm pleased about your fresh GUI and small footprint (go golang), but I have the feeling I should wait for some months to try ZITADEL again. When it's more matured. I don't have much experience regarding SAML and OIDC and never know, if I did a mistake or if somewhere a feature is missing (IDP or App).
Just to explain what I mean with "don't have much experience":
One day after my posting I found out that it's not only the missing mail address. Also logout doesn't work properly if the "team" feature is enabled in Grist. With Keycloak it works.
Then there is a 2nd app in my homelab that supports SAML (SOGo) and I also had some questions, because I don't find any option in ZITADEL.

**Reply by stebenz** [reply]
Hi @helmut72 

> It's a great convenience when it works in the wizard. When you re-enter the SAML configuration it's still there, but doesn't work. Don't you think it must be hidden or grayed out then, because otherwise it's an error?

IMO the metadata should be editable still, but it's maybe a little confusing that if you provide an URL that the metadata is currently not reloaded.

> No serious timeframe but I want to finish my private homelab in spring 2023. After a lot of research in the last two years. To continue my real project being based on my homelab. For now I just test out ZITADEL against other IDPs.

Ah, interesting, if you encounter any other problem just let me know.

> Don't know what features an IDP and what a SP must at least provide, that both work together.
> 
> I'm pleased about your fresh GUI and small footprint (go golang), but I have the feeling I should wait for some months to try ZITADEL again. When it's more matured. I don't have much experience regarding SAML and OIDC and never know, if I did a mistake or if somewhere a feature is missing (IDP or App). Just to explain what I mean with "don't have much experience": One day after my posting I found out that it's not only the missing mail address. Also logout doesn't work properly if the "team" feature is enabled in Grist. With Keycloak it works. Then there is a 2nd app in my homelab that supports SAML (SOGo) and I also had some questions, because I don't find any option in ZITADEL.

Good to hear that it seems interesting, we have to overthink the configuration for SAML at some point, as it seems that some applications which we didn't have on our radar have some trouble with the attribute names.
If I can help with any ZITADEL-related questions don't hesitate to ask, maybe our discord would maybe be a good spot to include some more people in the discussion.

**Comment by stebenz** [comment, marked as answer]
I will mark this replay as an answer, I hope I could clear up the remaining questions, if any more pop up, start a new discussion or a reply and I will try to help out.

**Reply by helmut72** [reply]
ok
